### PR TITLE
Performance: Improve Scheme.decode performance

### PR DESCRIPTION
**Pull request by ShrutiVerma97**
Benchmarks added in the PR.

Results:

Main Branch:
```
[info] Benchmark                                     Mode  Cnt    Score    Error  Units
[info] SchemeDecodeBenchmark.benchmarkSchemeDecode  thrpt    3  148.606 ? 12.130  ops/s
```

Current Branch:
```
[info] Benchmark                               Mode  Cnt    Score      Error  Units
[info] SchemeBenchmark.benchmarkSchemeDecode  thrpt    3  550.254 ? 2245.350  ops/s
```

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/http/Scheme.scala [hunk 1]
```diff
@@ -40,12 +40,12 @@ sealed trait Scheme { self =>
 }
 object Scheme       {
 
-  def decode(scheme: String): Option[Scheme] = scheme.toUpperCase match {
-    case "HTTPS" => Option(HTTPS)
-    case "HTTP"  => Option(HTTP)
-    case "WS"    => Option(WS)
-    case "WSS"   => Option(WSS)
-    case _       => None
+  private[zhttp] def unsafeDecode(scheme: String): Scheme = scheme.length match {
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/http/Scheme.scala]: this needs to rethink pattern matching on length will match on "XYZSD"

File: zio-http/src/test/scala/zhttp/http/SchemeSpec.scala [hunk 2]
```diff
@@ -9,7 +9,7 @@ object SchemeSpec extends DefaultRunnableSpec {
   override def spec = suite("SchemeSpec") {
     testM("string") {
       checkAll(HttpGen.scheme) { scheme =>
-        assertTrue(Scheme.decode(scheme.encode).get == scheme)
+        assertTrue(Scheme.unsafeDecode(scheme.encode) == scheme)
```
Comment (amitksingh1490) [inline zio-http/src/test/scala/zhttp/http/SchemeSpec.scala]: add a test for null Pointer exception

**Review by sumawa — APPROVED** [review]
Approving subjected to Amit's comment about unsafeDecode approach.

File: zio-http/src/main/scala/zhttp/http/Scheme.scala [hunk 3]
```diff
@@ -40,12 +40,28 @@ sealed trait Scheme { self =>
 }
 object Scheme       {
 
-  def decode(scheme: String): Option[Scheme] = scheme.toUpperCase match {
-    case "HTTPS" => Option(HTTPS)
-    case "HTTP"  => Option(HTTP)
-    case "WS"    => Option(WS)
-    case "WSS"   => Option(WSS)
-    case _       => None
+  def decode(scheme: String): Option[Scheme] = {
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/http/Scheme.scala]: Add ScalaDoc.

File: zio-http/src/main/scala/zhttp/http/Scheme.scala [hunk 4]
```diff
@@ -40,12 +40,32 @@ sealed trait Scheme { self =>
...
+      scheme.toUpperCase match {
+        case "HTTPS" => Option(HTTPS)
+        case "HTTP"  => Option(HTTP)
+        case "WS"    => Option(WS)
+        case "WSS"   => Option(WSS)
+        case _       => None
+      }
+  }
+
+  private[zhttp] def unsafeDecode(scheme: String): Scheme = {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Scheme.scala:50]: call `unsafeDecode` inside of `decode`

**Review by amitksingh1490 — APPROVED** [review]

File: zio-http-benchmarks/src/main/scala/zhttp.benchmarks/SchemeDecodeBenchmark.scala [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+@State(Scope.Thread)
+@BenchmarkMode(Array(Mode.Throughput))
+@OutputTimeUnit(TimeUnit.SECONDS)
+class SchemeDecodeBenchmark {
+  private val MAX = 1000000
+
+  @Benchmark
+  def benchmarkSchemeDecode(): Unit = {
+    (0 to MAX).foreach(_ => Scheme.unsafeDecode("HTTP"))
```
Comment (tusharmath) [inline zio-http-benchmarks/src/main/scala/zhttp.benchmarks/SchemeDecodeBenchmark.scala]:
```suggestion
    (0 to MAX).foreach(_ => Scheme.decode("HTTP"))
```

Let's compare the two implementations (old vs new).

File: zio-http/src/main/scala/zhttp/http/Scheme.scala [hunk 6]
```diff
@@ -40,12 +40,23 @@ sealed trait Scheme { self =>
...
-    case "WSS"   => Option(WSS)
-    case _       => None
+  /**
+   * Decodes a string to an Option of Scheme. Returns None in case of
+   * null/non-valid Scheme
+   */
+  def decode(scheme: String): Option[Scheme] =
+    Option(unsafeDecode(scheme))
+
+  private[zhttp] def unsafeDecode(scheme: String): Scheme = {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Scheme.scala:50]: Keep this method as `private` and drop the unit tests for this method.

File: zio-http/src/main/scala/zhttp/http/Scheme.scala [hunk 7]
```diff
@@ -40,12 +40,23 @@ sealed trait Scheme { self =>
...
-    case "WSS"   => Option(WSS)
-    case _       => None
+  /**
+   * Decodes a string to an Option of Scheme. Returns None in case of
+   * null/non-valid Scheme
+   */
+  def decode(scheme: String): Option[Scheme] =
+    Option(unsafeDecode(scheme))
+
+  private def unsafeDecode(scheme: String): Scheme = {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Scheme.scala]:
```suggestion
  private[zhttp] def unsafeDecode(scheme: String): Scheme = {
```

Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/http/Scheme.scala]: We also need to use this API in somewhere to see improvement's in performance. I think that should be `Request`

**Review by tusharmath — APPROVED** [review]
